L1Trigger/L1TGlobal/src/L1TGlobalUtil.cc

```diff
@@ -97,84 +99,99 @@ void l1t::L1TGlobalUtil::retrieveL1Setup(const edm::EventSetup& evSetup) {
-	 m_prescaleFactorsAlgoTrig = &m_initialPrescaleFactorsAlgoTrig;
-	 m_triggerMaskAlgoTrig = &m_initialTriggerMaskAlgoTrig;
-	 m_triggerMaskVetoAlgoTrig = &m_initialTriggerMaskVetoAlgoTrig;
-       }
+    if( !( m_readPrescalesFromFile ) or !(m_algorithmTriggersUnprescaled && m_algorithmTriggersUnmasked) ){
+      unsigned long long l1GtPfAlgoCacheID = evSetup.get<L1TGlobalPrescalesVetosRcd>().cacheIdentifier();
 
-       //Pick which set we are using
-       if(m_PreScaleColumn > m_prescaleFactorsAlgoTrig->size() || m_PreScaleColumn < 1) {	  
-	  LogTrace("l1t|Global")
-	   << "\nNo Prescale Set: " << m_PreScaleColumn
-	   << "\nMax Prescale Set value : " << m_prescaleFactorsAlgoTrig->size()  
-	    << "\nSetting prescale column to 1"
-	    << std::endl;
-	 m_PreScaleColumn = 1;
-       }
-       LogDebug("l1t|Global") << "Grabing prescale column "<< m_PreScaleColumn << endl;
-       const std::vector<int>& prescaleSet = (*m_prescaleFactorsAlgoTrig)[m_PreScaleColumn-1];
-       
-       // If masks or prescales enabled, get mask decision from prescale column
-       if( !( m_algorithmTriggersUnprescaled && m_algorithmTriggersUnmasked ) ){
-	 // For now, set masks according to prescale value of 0
-	 m_initialTriggerMaskAlgoTrig.clear();
-	 for( unsigned int iAlgo=0; iAlgo < prescaleSet.size(); iAlgo++ ){
-	   unsigned int value = prescaleSet[iAlgo];
-	   value = ( value==0 ) ? 0 : 1;
-	   m_initialTriggerMaskAlgoTrig.push_back(value);
-	 }
-	 m_triggerMaskAlgoTrig = &m_initialTriggerMaskAlgoTrig;
-       }
-               
-       for (std::map<std::string, L1TUtmAlgorithm>::const_iterator itAlgo = m_algorithmMap->begin(); itAlgo != m_algorithmMap->end(); itAlgo++) {
+      if (m_l1GtPfAlgoCacheID != l1GtPfAlgoCacheID) {
+	//std::cout << "Reading Prescales and Masks from dB" << std::endl;
 
-          // Get the algorithm name
-          std::string algName = itAlgo->first;
-          int algBit = (itAlgo->second).getIndex(); //algoBitNumber();
+	// clear and dimension
+	resetPrescaleVectors();
+	resetMaskVectors();
 
-	  (m_prescales[algBit]).first  = algName;
-	  (m_prescales[algBit]).second = prescaleSet[algBit];
+	edm::ESHandle< L1TGlobalPrescalesVetos > l1GtPrescalesVetoes;
+	evSetup.get< L1TGlobalPrescalesVetosRcd >().get( l1GtPrescalesVetoes );
+	const L1TGlobalPrescalesVetos * es = l1GtPrescalesVetoes.product();
+	m_l1GtPrescalesVetoes = PrescalesVetosHelper::readFromEventSetup(es);
 
-	  (m_masks[algBit]).first  = algName;
-	  (m_masks[algBit]).second = (*m_triggerMaskAlgoTrig)[algBit];	  
+	m_prescaleFactorsAlgoTrig = &(m_l1GtPrescalesVetoes->prescaleTable());
+	m_numberPhysTriggers = (*m_prescaleFactorsAlgoTrig)[0].size(); // assumes all prescale columns are the same length
 
-	  (m_vetoMasks[algBit]).first  = algName;
-	  (m_vetoMasks[algBit]).second = (*m_triggerMaskVetoAlgoTrig)[algBit];
-       }
-       
-      m_filledPrescales = true;
+	m_triggerMaskAlgoTrig   = &(m_l1GtPrescalesVetoes->triggerAlgoBxMask());
```

Comment:
so triggerAlgoBxMask() is already returning the right thing - a const ref. why it not be used directly?